src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2716,12 +2716,20 @@ void ProcessRequiresUnreferencedCode (MethodDefinition method, DependencyKind de
 
 		internal void CheckAndReportRequiresUnreferencedCode (MethodDefinition method)
 		{
-			var currentScope = _scopeStack.CurrentScope;
+			var currentOrigin = _scopeStack.CurrentScope.Origin;
 
 			// If the caller of a method is already marked with `RequiresUnreferencedCodeAttribute` a new warning should not
 			// be produced for the callee.
-			if (currentScope.Origin.MemberDefinition != null &&
-				Annotations.HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (currentScope.Origin.MemberDefinition))
+			// Check both the origin's member (the direct method which contains the call to RUC)
+			// as well as the suppression context member (the "user code" source in case of compiler generated code)
+			IMemberDefinition suppressionContextMember = currentOrigin.SuppressionContextMember;
+			if (suppressionContextMember != null &&
+				Annotations.HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (suppressionContextMember))
+				return;
+
+			IMemberDefinition originMember = currentOrigin.MemberDefinition;
+			if (suppressionContextMember != originMember && originMember != null &&
+				Annotations.HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (originMember))
```

Comment:
Can these two checks be extracted into a helper and shared with https://github.com/mono/linker/pull/2075/files#diff-565c4d3a1fda40fe49e36a001479e217a6a484722f6e97ee132c3ddcefdbd585R64?